cln_plugin/server.go

```diff
@@ -0,0 +1,220 @@
+package cln_plugin
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	grpc "google.golang.org/grpc"
+)
+
+var receiveWaitDelay = time.Millisecond * 200
+
+type subscription struct {
+	stream ClnPlugin_HtlcStreamServer
+	done   chan struct{}
+}
+type server struct {
+	ClnPluginServer
+	listenAddress string
+	grpcServer    *grpc.Server
+	startMtx      sync.Mutex
+	corrMtx       sync.Mutex
+	subscription  *subscription
+	newSubscriber chan struct{}
+	done          chan struct{}
+	correlations  map[uint64]chan *HtlcResolution
+	index         uint64
+}
+
+func NewServer(listenAddress string) *server {
+	return &server{
+		listenAddress: listenAddress,
+		newSubscriber: make(chan struct{}, 1),
+		correlations:  make(map[uint64]chan *HtlcResolution),
+		index:         0,
+	}
+}
+
+func (s *server) Start() error {
+	s.startMtx.Lock()
+	if s.grpcServer != nil {
+		s.startMtx.Unlock()
+		return nil
+	}
+
+	lis, err := net.Listen("tcp", s.listenAddress)
+	if err != nil {
+		log.Printf("ERROR Server failed to listen: %v", err)
+		s.startMtx.Unlock()
+		return err
+	}
+
+	s.done = make(chan struct{})
+	s.grpcServer = grpc.NewServer()
+	s.startMtx.Unlock()
+	RegisterClnPluginServer(s.grpcServer, s)
+
+	log.Printf("Server starting to listen on %s.", s.listenAddress)
+	go s.listenHtlcResponses()
+	return s.grpcServer.Serve(lis)
+}
+
+func (s *server) Stop() {
+	s.startMtx.Lock()
+	defer s.startMtx.Unlock()
+	log.Printf("Server Stop() called.")
+	if s.grpcServer == nil {
+		return
+	}
+
+	close(s.done)
+	s.grpcServer.Stop()
+	s.grpcServer = nil
+}
+
+func (s *server) HtlcStream(stream ClnPlugin_HtlcStreamServer) error {
+	log.Printf("Got HTLC stream subscription request.")
+	s.startMtx.Lock()
+	if s.subscription != nil {
+		s.startMtx.Unlock()
+		return fmt.Errorf("already subscribed")
+	}
+
+	sb := &subscription{
+		stream: stream,
+		done:   make(chan struct{}),
+	}
+	s.subscription = sb
+	s.newSubscriber <- struct{}{}
+	s.startMtx.Unlock()
+
+	defer func() {
+		s.startMtx.Lock()
+		s.subscription = nil
+		close(sb.done)
+		s.startMtx.Unlock()
+	}()
+
+	go func() {
+		<-stream.Context().Done()
+		log.Printf("HtlcStream context is done. Removing subscriber: %v", stream.Context().Err())
+		s.startMtx.Lock()
+		s.subscription = nil
+		close(sb.done)
+		s.startMtx.Unlock()
+	}()
+
+	select {
```

Comment:
Do we need a mechanism at the end that go over all held htlc in our internal `correlations` map and execute the default resolution for them? Otherwise they would held forever.

Reply:
In theory the 'send' method should do that.
It 
- sets the correlation id in the map (with a chan for waiting for the response)
- sends the htlc_accepted to lspd
- wait for the response, or done.
- if lspd does not return and the cln plugin quits (done), the correlation should be deleted from the correlations by the defer statement here https://github.com/breez/lspd/pull/42/files#diff-e6eb05c8c6fac8eb4b3353b7b77fcdfe875a0cf57db0b7b47f0800b20521ef96R137-R142

There are probably some cases where this flow will not work yet:
- when lspd stops without sending a response the plugin may wait until it's killed before resolving. I haven't yet built in a mechanism for handling lspd dropping out. I'm not sure whether the receive side and the send side are handled independently by grpc. So if the send side drops, should we close the receive side and vice versa?
- I do see now we're not currently waiting for the correlations map to clear in the `Stop` method on the server. And I see the plugin is now stopped before the server, that should be the other way around.
- definately worth testing these flows.